src/Estimators/StructureFactorEstimator.h

```diff
@@ -0,0 +1,100 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2024 QMCPACK developers.
+//
+// File developed by: Peter W. Doak, doakpw@ornl.gov, Oak Ridge National Laboratory
+//
+// Some code refactored from: QMCHamiltonian/{SkEstimator.h, SkAllEstimator.h}
+//////////////////////////////////////////////////////////////////////////////////////
+
+#ifndef QMCPLUSPLUS_STRUCTUREFACTORESTIMATOR_H
+#define QMCPLUSPLUS_STRUCTUREFACTORESTIMATOR_H
+
+#include "OperatorEstBase.h"
+#include "StructureFactorInput.h"
+
+namespace qmcplusplus
+{
+
+namespace testing
+{
+class StructureFactorAccess;
+}
+
+class StructureFactorEstimator : public OperatorEstBase
+{
+public:
+  using QMCT         = QMCTraits;
+  using Real         = QMCT::RealType;
+  using FullPrecReal = QMCT::FullPrecRealType;
+  using KPt          = TinyVector<Real, QMCTraits::DIM>;
+
+  StructureFactorEstimator(StructureFactorInput& sfi,
+                           ParticleSet& pset_ions,
+                           ParticleSet& pset_elec,
+                           DataLocality data_locality = DataLocality::crowd);
+
+  /** accumulate 1 or more walkers of EnergyDensity samples
+   */
+  void accumulate(const RefVector<MCPWalker>& walkers,
+                  const RefVector<ParticleSet>& psets,
+                  const RefVector<TrialWaveFunction>& wfns,
+                  const RefVector<QMCHamiltonian>& hams,
+                  RandomBase<FullPrecReal>& rng) override;
+
+  /** start block entry point
+   */
+  void startBlock(int steps) override;
+
+  UPtr<OperatorEstBase> spawnCrowdClone() const override;
+
+  void registerOperatorEstimator(hdf_archive& file) override;
+  void write(hdf_archive& file);
+  void collect(const RefVector<OperatorEstBase>& type_erased_operator_estimators) override;
+
+  long long getNumKPoints() { return num_kpoints_; }
+protected:
+  // Testing functions
+  const Vector<Real>& getSKElecElec() const { return sfk_e_e_; }
+  const Vector<std::complex<Real>>& getRhoKElec() const { return rhok_e_; }
+
+private:
+  int elec_species_;
+  ParticleSet& elns_;
+  int ion_species_;
+  ParticleSet& ions_;
+
+  /// number of k points
+  long long num_kpoints_;
```

Comment:
Could you rename variables to `elec_num_species_` and `ion_num_species_`? The current names sound like array of something. Could you make all the 5 entries const?

Reply:
sorry. Just noticed that `num_kpoints_` also can be const as well as `getNumKPoints()`